mailbox/backup/src/main/java/org/apache/james/mailbox/backup/DefaultMailboxBackup.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.james.mailbox.store.mail.model.Mailbox;
+import org.apache.james.util.OptionalUtils;
+import org.apache.james.util.streams.Iterators;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.steveash.guavate.Guavate;
+
+public class DefaultMailboxBackup implements MailboxBackup {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultMailboxBackup.class);
+
+    protected class MailAccountContent {
+        final MailboxWithAnnotations mailboxWithAnnotations;
+        final Stream<MessageResult> messages;
+
+        MailAccountContent(MailboxWithAnnotations mailboxWithAnnotations, Stream<MessageResult> messages) {
+            this.mailboxWithAnnotations = mailboxWithAnnotations;
+            this.messages = messages;
+        }
+    }
+
+    public DefaultMailboxBackup(MailboxManager mailboxManager, ArchiveService archiveService, BlobStore store) {
+        this.mailboxManager = mailboxManager;
+        this.archiveService = archiveService;
+        this.store = store;
+    }
+
+    private final MailboxManager mailboxManager;
+    private final ArchiveService archiveService;
+    private final BlobStore store;
+
+    private Function<MailboxPath, Optional<MailAccountContent>> getMailboxWithAnnotationsFromPath(MailboxSession session) {
+        return path -> {
+            try {
+                StoreMessageManager messageManager = (StoreMessageManager) mailboxManager.getMailbox(path, session);
+                Mailbox mailbox = messageManager.getMailboxEntity();
+                List<MailboxAnnotation> annotations = mailboxManager.getAllAnnotations(path, session);
+                MailboxWithAnnotations mailboxWithAnnotations = new MailboxWithAnnotations(mailbox, annotations);
+                Stream<MessageResult> messages = Iterators.toStream(messageManager.getMessages(MessageRange.all(), FetchGroupImpl.FULL_CONTENT, session));
+                return Optional.of(new MailAccountContent(mailboxWithAnnotations, messages));
+            } catch (MailboxException e) {
+                LOGGER.error("Error while fetching Mailbox during backup", e);
+                return Optional.empty();
+            }
+        };
+    }
+
+    private List<MailAccountContent> getAccountContentForUser(MailboxSession session) throws MailboxException {
+        MailboxQuery queryUser = MailboxQuery.builder().username(session.getUser().asString()).build();
+        Stream<MailboxPath> paths = mailboxManager.search(queryUser, session).stream().map(mailboxMetaData -> mailboxMetaData.getPath());
+        List<MailAccountContent> mailboxes = paths
+            .flatMap(getMailboxWithAnnotationsFromPath(session).andThen(OptionalUtils::toStream))
+            .collect(Guavate.toImmutableList());
+
+        return mailboxes;
+    }
+
+    private Publisher<BlobId> saveToStore(List<MailboxWithAnnotations> mailboxes, Stream<MessageResult> messages) throws IOException {
```

Comment:
Don't we want to be using a piped inputstream/outputStream for this?

CF the cool work of @trantienduchn on https://github.com/linagora/james-project/pull/2248